### PR TITLE
SSAO

### PR DESCRIPTION
**Pull request by ffreyer**
Based on the learnopengl tutorial Simon posted in https://github.com/JuliaPlots/Makie.jl/issues/487

I believe SSAO is fully working at this point, but it's a very hacky implementation. I thought this would be a good point to get some feedback though, specifically on how connect this to GLMakie correctly.

Rough TODO
- [x] Pass `projection` in a less hacky way
- [x] Make things controllable using Makie attributes. (How would I go about this? They aren't directly available when creating the Framebuffer...) This includes:
  * ~~number of SSAO samples~~ (requires shader recompilation - how?)
  * bias (minimum depth difference for pixel to be occluded)
  * radius (distance checked for occlusion)
  * blur (related to number of pixels used in blur)
  * ~~maybe some controls for randomness of noise and samples~~
- [x] on/off switch for SSAO
- ~~implement deferred shading (i.e. move light calculations to postprocessing) (not doing this results in a weird glowing effect)~~ (nvm, the glow is a result of blurring)
- [x] fix background getting darkened (when there is no geometry)
-  ~~allow SSAO to run w/o FXAA~~
- [x] fix lighting
- [x] fix glow
- [x] replace `render_colorbuffer`
---

Here are some pics:
cat w/o SSAO:
noSSAO

cat w/ SSAO
withSSAO

cat w/ only SSAO
occlusion

### CODE REVIEW

File: src/GLVisualize/assets/shader/util.vert [hunk 1]
```diff
@@ -228,21 +228,32 @@ vec4 _color(Nothing color, float intensity, sampler1D color_map, vec2 color_norm
...
     o_lightdir             = normalize(lightposition - position_world.xyz);
     // direction to camera
     o_camdir               = normalize(eyeposition - position_world.xyz);
+    // Shouldn't this just be position_world?
+    o_view_pos             = view * position_world;
     // screen space coordinates of the vertex
-    gl_Position            = projection * view * position_world;
+    gl_Position            = projection * o_view_pos;
+    // TODO calculate w/ normalmatrix (see o_normal links)
+    o_view_normal          = transpose(inverse(mat3(view))) * normal;
```
Comment (SimonDanisch) [inline src/GLVisualize/assets/shader/util.vert:256]: Yeah we shouldn't calculate this in the shader  ;)

**Comment by SimonDanisch** [comment]
Awesome :) On first sight, this looks pretty similar to how I'd have done it ;)

**Comment by ffreyer** [comment]
I think I got SSAO working correctly for multiple scenes now. Took me quite a while to figure out why it wasn't working. I wrote a longer comment in the shader that explains what I'm doing and why I'm doing it.

Current cat(s):

Screenshot from 2020-04-30 04-12-22

File: src/rendering.jl [hunk 2]
```diff
@@ -82,32 +87,53 @@ function render_frame(screen::Screen)
...
+
+    # FXAA - calculate LUMA
+    glBindFramebuffer(GL_FRAMEBUFFER, fb.id[2])
+    glDrawBuffer(GL_COLOR_ATTACHMENT0)  # color_luma buffer
     glViewport(0, 0, w, h)
-    glDrawBuffer(GL_COLOR_ATTACHMENT0)
-    GLAbstraction.render(fb.postprocess[2]) # copy with fxaa postprocess
+    # clear shouldn't be necessary. every pixel should be written to
+    glClearColor(1, 0, 1, 1)
+    glClear(GL_COLOR_BUFFER_BIT)
```
Comment (ffreyer) [inline src/rendering.jl:112]: I made this clear color a jarring pink so I can see if it ever appears. Should be removed or set to white before merging. There's another one of these further down (using green)

Comment (ffreyer) [reply src/rendering.jl]: Becomes visible if `bias < 0`.

File: src/GLVisualize/assets/shader/util.vert (hunk 1 above)
Comment (ffreyer) [reply src/GLVisualize/assets/shader/util.vert]: I added some code to generate a normalmatrix in `cached_robj!()`

File: src/GLVisualize/assets/shader/util.vert [hunk 3]
```diff
@@ -225,24 +225,27 @@ vec4 _color(Nothing color, float intensity, sampler1D color_map, vec2 color_norm
...
     // direction to light
     o_lightdir             = normalize(lightposition - position_world.xyz);
     // direction to camera
     o_camdir               = normalize(eyeposition - position_world.xyz);
-    // screen space coordinates of the vertex
-    gl_Position            = projection * view * position_world;
+    // position in view space (as seen from camera)
+    o_view_pos             = view * position_world;
+    // position in clip space (w/ depth)
+    gl_Position            = projection * o_view_pos;
```
Comment (ffreyer) [inline src/GLVisualize/assets/shader/util.vert:248]: Do these require a `v /= v.w`?

File: src/GLVisualize/assets/shader/util.vert [hunk 4]
```diff
@@ -225,24 +219,29 @@ vec4 _color(Nothing color, float intensity, sampler1D color_map, vec2 color_norm
...
     // direction to light
-    o_lightdir = normalize(lightposition - position_world.xyz);
+    o_lightdir             = normalize(view*vec4(lightposition, 1.0) - o_view_pos).xyz;
     // direction to camera
-    o_camdir = normalize(eyeposition - position_world.xyz);
-    // screen space coordinates of the vertex
-    gl_Position = projection * view * position_world;
+    o_camdir               = normalize(view*vec4(eyeposition, 1.0) - o_view_pos).xyz;
+    // position in clip space (w/ depth)
+    gl_Position            = projection * o_view_pos;
```
Comment (ffreyer) [inline src/GLVisualize/assets/shader/util.vert:244]: Alternatively, we could also do
```GLSL
o_normal = normal_model_matrix * normal;
o_view_normal = normal_view_matrix * normal; // pass this to SSAO
o_lightdir = normalize(lightposition - position_world.xyz);
o_camdir = normalize(eyeposition - position_world.xyz);
o_view_pos = view * position_world; // pass this to SSAO
gl_Position = projection * o_view_pos;
```
where the normal matrices are `transpose(inv(view)` and `tranpose(inv(model))`. Not sure which is better...

**Comment by ffreyer** [comment]
```julia
using Colors, Makie
s1, radius = textslider(0.0f0:0.1f0:2f0, "Radius", start = 0.5f0)
s2, bias = textslider(0f0:0.005f0:0.1f0, "Bias", start = 0.025f0)
s3, blur = textslider(Int32(0):Int32(1):Int32(5), "Blur", start = Int32(2))
ssao_attrib = Attributes(radius=radius, bias=bias, blur=blur)

ps = rand(Point3f0, 100)
colors = rand(RGB, 100)
scene1 = meshscatter(ps, color=colors, ssao=true, SSAO=ssao_attrib)
scene2 = meshscatter(ps, color=colors, ssao=false) # default
hbox(vbox(s1, s2, s3), vbox(scene1, scene2))
```
Here's a little example for playing around with attributes. The left side uses SSAO, the right side doesn't. Depending on the colors the shine/glow issue is quite noticeable here, and by modifying `blur` (which sets the pixel-range for blurring) you can see that it is connected to that.

**Pull request merged**